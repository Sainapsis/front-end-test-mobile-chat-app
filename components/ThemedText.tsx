import React from 'react';
import { Text, TextStyle, TextProps, useColorScheme } from 'react-native';
import { Colors } from '@/constants/Colors';

interface ThemedTextProps extends TextProps {
  readonly children: React.ReactNode;
  readonly style?: TextStyle;
  readonly type?: 'title' | 'subtitle' | 'body' | 'caption' | 'defaultSemiBold' | 'default' | 'link';
}

export function ThemedText({ children, style, type = 'body', ...props }: ThemedTextProps) {
  const colorScheme = useColorScheme();
  const theme = colorScheme === 'dark' ? Colors.dark : Colors.light;

  const textStyles = {
    title: {
      fontSize: 24,
      fontWeight: 'bold',
      color: theme.text,
    },
    subtitle: {
      fontSize: 18,
      fontWeight: '600',
      color: theme.text,
    },
    body: {
      fontSize: 16,
      color: theme.text,
    },
    caption: {
      fontSize: 14,
      color: theme.text,
    },
    default: {
      fontSize: 16,
      color: theme.text,
    },
    defaultSemiBold: {
      fontSize: 16,
      fontWeight: '600',
      color: theme.text,
    },
    link: {
      fontSize: 16,
      color: theme.primary,
      textDecorationLine: 'underline',
    }
  };

  return (
    <Text style={[textStyles[type], style]} {...props}>
      {children}
    </Text>
  );
}
