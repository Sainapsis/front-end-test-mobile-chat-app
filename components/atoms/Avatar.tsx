// TP
import { View, StyleSheet } from "react-native";
import React from "react";
// BL
import { UserInterface } from "@/interfaces/User.interface";

// UI
import { ThemedText } from "../ThemedText";

interface AvatarProps {
  user?: UserInterface;
  size?: number;
  showStatus?: boolean;
}

const getAvatarColor = (identifier?: string): string => {
  if (!identifier) return "#C0C0C0";

  let hash = 0;
  for (let i = 0; i < identifier.length; i++) {
    hash = identifier.charCodeAt(i) + ((hash << 5) - hash);
  }

  let color = "#";
  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += ("00" + value.toString(16)).substr(-2);
  }

  return color;
};

const getInitials = (name?: string): string => {
  if (!name) return "?";

  const names = name.split(" ");
  if (names.length === 1) {
    return names[0].charAt(0).toUpperCase();
  }

  return (names[0].charAt(0) + names[names.length - 1].charAt(0)).toUpperCase();
};

export function Avatar({ user, size = 40, showStatus = true }: AvatarProps) {
  const backgroundColor = getAvatarColor(user?.id || user?.name);
  const initials = getInitials(user?.name);

  const statusColors = {
    online: "#4CAF50",
    offline: "#9E9E9E",
    away: "#FFC107",
  };

  return (
    <View style={styles.container}>
      <View
        style={[
          styles.avatar,
          {
            width: size,
            height: size,
            borderRadius: size / 2,
            backgroundColor,
          },
        ]}
      >
        <ThemedText style={[styles.initials, { fontSize: size * 0.4 }]}>
          {initials}
        </ThemedText>
      </View>
      {showStatus && user?.status && (
        <View
          style={[
            styles.statusIndicator,
            {
              backgroundColor: statusColors[user.status],
              width: size / 4,
              height: size / 4,
              borderRadius: size / 8,
              right: 0,
              bottom: 0,
            },
          ]}
        />
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    position: "relative",
  },
  avatar: {
    alignItems: "center",
    justifyContent: "center",
  },
  initials: {
    color: "white",
    fontWeight: "bold",
  },
  statusIndicator: {
    position: "absolute",
    borderWidth: 1.5,
    borderColor: "white",
  },
});
