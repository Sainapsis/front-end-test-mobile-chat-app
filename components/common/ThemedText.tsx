import { Text, type TextProps } from 'react-native';
import { themedTextStyles } from '@/styles/common/themedText.styles';
import { useThemeColor } from '@/hooks/useThemeColor';

export type ThemedTextProps = TextProps & {
  lightColor?: string;
  darkColor?: string;
  type?: 'default' | 'title' | 'defaultSemiBold' | 'subtitle' | 'link';
};

export function ThemedText({
  style,
  lightColor,
  darkColor,
  type = 'default',
  ...rest
}: ThemedTextProps) {
  const color = useThemeColor({ light: lightColor, dark: darkColor }, 'text');

  return (
    <Text
      style={[
        { color },
        type === 'default' ? themedTextStyles.default : undefined,
        type === 'title' ? themedTextStyles.title : undefined,
        type === 'defaultSemiBold' ? themedTextStyles.defaultSemiBold : undefined,
        type === 'subtitle' ? themedTextStyles.subtitle : undefined,
        type === 'link' ? themedTextStyles.link : undefined,
        style,
      ]}
      {...rest}
    />
  );
}


