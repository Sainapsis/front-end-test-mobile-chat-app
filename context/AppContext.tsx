import React, { createContext, useContext, ReactNode } from 'react';
import { useUser } from '@/hooks/useUser';
import { useChats } from '@/hooks/useChats';
import { DatabaseProvider } from '../database/DatabaseProvider';
import { useDatabase } from '@/hooks/useDatabase';
import { User } from '@/types/User';
import { Chat } from '@/types/Chat';

type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (chatId: string, text: string, senderId: string) => Promise<boolean>;
  loading: boolean;
  dbInitialized: boolean;
  clearChats: (userId: string) => Promise<void>;
  deleteChat?: (chatId: string, userId: string) => Promise<void>;
  deleteMessage?: (messageId: string, chatId: string) => Promise<boolean>;
  addReaction?: (messageId: string, emoji: string) => Promise<boolean>;
  removeReaction?: (reactionId: string, messageId: string) => Promise<boolean>;
  editMessage?: (messageId: string, newText: string) => Promise<boolean>;
};

const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  const chatContext = useChats(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value: AppContextType = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
    deleteChat: chatContext.deleteChat || undefined,
    deleteMessage: chatContext.deleteMessage || undefined,
    addReaction: chatContext.addReaction || undefined,
    removeReaction: chatContext.removeReaction || undefined,
    editMessage: chatContext.editMessage || undefined,
  };

  if (!isInitialized) {
    return null;
  }

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export const AppProvider = ({ children }: { children: React.ReactNode }) => (
  <DatabaseProvider>
    <AppContent>{children}</AppContent>
  </DatabaseProvider>
);

export function useAppContext() {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
}
