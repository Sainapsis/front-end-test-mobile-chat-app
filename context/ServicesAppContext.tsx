import React, { createContext, useContext, ReactNode } from 'react';
import { DatabaseProvider, useDatabaseStatus } from './DataBaseContext';
import { useChatsDb,  useUserDb } from '../hooks/db';
import { Chat, Message, User } from '@/utils/types';

type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (
    chatId: string, 
    text: string | undefined, 
    senderId: string,
    mediaUrl?: string,
    mediaType?: string,
    mediaThumbnail?: string,
    voiceUrl?: string,
    voiceDuration?: number,
    isVoiceMessage?: boolean,
    
  ) => Promise<Message | null>;
  addReaction: (messageId: string, userId: string, emoji: string) => Promise<boolean>;
  removeReaction: (reactionId: string) => Promise<boolean>;
  searchMessages: (query: string, chatId?: string) => Promise<void>;
  searchResults: Message[];
  deleteMessage: (messageId: string) => Promise<boolean>;
  editMessage: (messageId: string, newText: string) => Promise<boolean>;
  updateMessageStatus: (messageId: string, status: 'sent' | 'delivered' | 'read') => Promise<void>;
  markMessagesAsRead: (chatId: string) => Promise<void>;
  loading: boolean;
  dbInitialized: boolean;
};

export const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabaseStatus();
  const userContext = useUserDb();
  const chatContext = useChatsDb(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ children }: { children: ReactNode }) {
  return (
    <DatabaseProvider>
      <AppContent>{children}</AppContent>
    </DatabaseProvider>
  );
}

