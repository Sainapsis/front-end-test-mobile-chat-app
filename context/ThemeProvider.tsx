import React, { useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import ThemeContext from './ThemeContext';
import { Theme } from '@/types/tColores';

/**
 * Provider component that handles theme state and persistence
 * @param children - Child components to be wrapped by the provider
 */
export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>('light');

  useEffect(() => {
    loadTheme();
  }, []);

  const loadTheme = async () => {
    try {
      const savedTheme = await AsyncStorage.getItem('theme');
      if (savedTheme) {
        setTheme(savedTheme as Theme);
      }
    } catch (error) {
      console.error('Error loading theme:', error);
    }
  };

  const toggleTheme = async () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    try {
      await AsyncStorage.setItem('theme', newTheme);
    } catch (error) {
      console.error('Error saving theme:', error);
    }
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};