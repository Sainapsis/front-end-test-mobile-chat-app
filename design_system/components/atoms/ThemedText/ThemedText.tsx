import { Text, type TextProps } from 'react-native';
import { useThemeColor } from '@/hooks/useThemeColor';
import { themes } from '@/design_system/ui/tokens';
import { styles } from './ThemedText.styles';

export type ThemedTextProps = TextProps & {
  /** Color to be used in light mode */
  lightColor?: string;
  /** Color to be used in dark mode */
  darkColor?: string;
  /** Variant of the text color based on theme */
  textVariant?: keyof typeof themes.light.text;
  /** Type of text style to apply */
  type?: 'default' | 'title' | 'defaultSemiBold' | 'subtitle' | 'link';
};

/**
 * ThemedText component applies theme-based colors and styles to text.
 * It supports different color variants and text types for customization.
 */
export function ThemedText({
  style,
  lightColor,
  darkColor,
  textVariant = 'primary',
  type = 'default',
  ...rest
}: ThemedTextProps) {
  const color = useThemeColor(
    { light: lightColor, dark: darkColor },
    'text',
    textVariant
  );

  return (
    <Text
      style={[
        { color },
        styles[type],
        style, // Maintains custom styles
      ]}
      {...rest}
    />
  );
}
