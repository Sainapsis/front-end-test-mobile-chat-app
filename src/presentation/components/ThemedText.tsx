import { useThemeColor } from '@/src/presentation/hooks/useThemeColor';
import { Text, type TextProps, StyleSheet } from 'react-native';


export enum TextType {
  DEFAULT = 'default',
  TITLE = 'title',
  DEFAULT_SEMI_BOLD = 'defaultSemiBold',
  SUBTITLE = 'subtitle',
  LINK = 'link',
}

export type ThemedTextProps = TextProps & {
  lightColor?: string;
  darkColor?: string;
  type?: TextType;
};

export function ThemedText({
  style,
  lightColor,
  darkColor,
  type = TextType.DEFAULT,
  ...rest
}: ThemedTextProps) {
  const color = useThemeColor({ light: lightColor, dark: darkColor }, 'text');

  return (
    <Text
      style={[
        { color },
        type === TextType.DEFAULT ? styles.default : undefined,
        type === TextType.TITLE ? styles.title : undefined,
        type === TextType.DEFAULT_SEMI_BOLD ? styles.defaultSemiBold : undefined,
        type === TextType.SUBTITLE ? styles.subtitle : undefined,
        type === TextType.LINK ? styles.link : undefined,
        style,
      ]}
      {...rest}
    />
  );
}

const styles = StyleSheet.create({
  default: {
    fontSize: 16,
    lineHeight: 24,
  },
  defaultSemiBold: {
    fontSize: 16,
    lineHeight: 24,
    fontWeight: '600',
  },
  title: {
    fontSize: 32,
    fontWeight: 'bold',
    lineHeight: 32,
  },
  subtitle: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  link: {
    lineHeight: 30,
    fontSize: 16,
    color: '#0a7ea4',
  },
});
