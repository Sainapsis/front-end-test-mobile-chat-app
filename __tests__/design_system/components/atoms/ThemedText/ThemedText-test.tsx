import * as React from 'react';
import { render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@/context/ThemeContext';
import { ThemedText } from '@/design_system/components/atoms/ThemedText';

describe('ThemedText Component', () => {
  const renderWithTheme = (component: React.ReactElement) => {
    return render(
      <ThemeProvider>
        {component}
      </ThemeProvider>
    );
  };

  it('renders correctly with snapshot', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <ThemedText>Snapshot test!</ThemedText>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders text content correctly', () => {
    const testText = 'Hello World';
    const { getByText } = renderWithTheme(<ThemedText>{testText}</ThemedText>);
    expect(getByText(testText)).toBeTruthy();
  });

  it('applies custom styles when provided', () => {
    const customStyle = { fontSize: 20, fontWeight: 'bold' as const };
    const { toJSON } = renderWithTheme(<ThemedText style={customStyle}>Styled Text</ThemedText>);
    expect(toJSON()).not.toBeNull();
  });

  it('renders with different text variants', () => {
    const textVariants = ['primary', 'secondary', 'disabled', 'contrast', 'black'];
    
    textVariants.forEach(variant => {
      const { getByText } = renderWithTheme(
        <ThemedText textVariant={variant as any}>{`${variant} text`}</ThemedText>
      );
      expect(getByText(`${variant} text`)).toBeTruthy();
    });
  });

  it('renders with different type variants', () => {
    const typeVariants = ['link', 'h1', 'h2', 'h3', 'body', 'caption'];
    
    typeVariants.forEach(type => {
      const { getByText } = renderWithTheme(
        <ThemedText type={type as any}>{`${type} text`}</ThemedText>
      );
      expect(getByText(`${type} text`)).toBeTruthy();
    });
  });

  it('handles onPress events', () => {
    const onPressMock = jest.fn();
    const { getByText } = renderWithTheme(
      <ThemedText onPress={onPressMock}>Clickable Text</ThemedText>
    );
    
    const textElement = getByText('Clickable Text');
    textElement.props.onPress();
    
    expect(onPressMock).toHaveBeenCalledTimes(1);
  });
});