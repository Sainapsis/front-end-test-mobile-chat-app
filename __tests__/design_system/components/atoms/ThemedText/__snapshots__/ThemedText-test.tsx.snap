// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedText Component renders correctly with snapshot 1`] = `
<Text
  style={
    [
      {
        "color": "#11181C",
      },
      {
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 24,
      },
      undefined,
    ]
  }
>
  Snapshot test!
</Text>
`;
